privval/signer_dialer_endpoint.go

```diff
@@ -0,0 +1,244 @@
+
+// SignerDialerEndpointTimeoutReadWrite sets the read and write timeout for connections
+// from external signing processes.
+func SignerDialerEndpointTimeoutReadWrite(timeout time.Duration) SignerServiceEndpointOption {
+	return func(ss *SignerDialerEndpoint) { ss.timeoutReadWrite = timeout }
+}
+
+// SignerDialerEndpointConnRetries sets the amount of attempted retries to AcceptNewConnection.
+func SignerDialerEndpointConnRetries(retries int) SignerServiceEndpointOption {
+	return func(ss *SignerDialerEndpoint) { ss.maxConnRetries = retries }
+}
+
+// SignerDialerEndpoint dials using its dialer and responds to any
+// signature requests using its privVal.
+type SignerDialerEndpoint struct {
+	cmn.BaseService
+
+	mtx    sync.Mutex
+	dialer SocketDialer
+	conn   net.Conn
+
+	timeoutReadWrite time.Duration
+	retryWait        time.Duration
+	maxConnRetries   int
+
+	chainID string
+	privVal types.PrivValidator
+
+	stopCh    chan struct{}
+	stoppedCh chan struct{}
+}
+
+// NewSignerDialerEndpoint returns a SignerDialerEndpoint that will dial using the given
+// dialer and respond to any signature requests over the connection
+// using the given privVal.
+func NewSignerDialerEndpoint(
+	logger log.Logger,
+	chainID string,
+	privVal types.PrivValidator,
+	dialer SocketDialer,
+) *SignerDialerEndpoint {
+
+	se := &SignerDialerEndpoint{
+		dialer:           dialer,
+		timeoutReadWrite: defaultTimeoutReadWriteSeconds * time.Second,
+		retryWait:        defaultRetryWaitMilliseconds * time.Millisecond,
+		maxConnRetries:   defaultMaxDialRetries,
+
+		chainID: chainID,
+		privVal: privVal,
+	}
+
+	se.BaseService = *cmn.NewBaseService(logger, "SignerDialerEndpoint", se)
+	return se
+}
+
+// OnStart implements cmn.Service.
+func (sd *SignerDialerEndpoint) OnStart() error {
+	sd.Logger.Debug("SignerDialerEndpoint: OnStart")
+
+	sd.stopCh = make(chan struct{})
+	sd.stoppedCh = make(chan struct{})
+
+	go sd.serviceLoop()
+
+	return nil
+}
+
+// OnStop implements cmn.Service.
+func (sd *SignerDialerEndpoint) OnStop() {
+	sd.Logger.Debug("SignerDialerEndpoint: OnStop calling Close")
+
+	// Stop service loop
+	close(sd.stopCh)
+	<-sd.stoppedCh
+
+	_ = sd.Close()
+}
+
+// Close closes the underlying net.Conn.
+func (sd *SignerDialerEndpoint) Close() error {
+	sd.mtx.Lock()
+	defer sd.mtx.Unlock()
+	sd.Logger.Debug("SignerDialerEndpoint: Close")
+
+	if sd.conn != nil {
+		if err := sd.conn.Close(); err != nil {
+			sd.Logger.Error("OnStop", "err", cmn.ErrorWrap(err, "closing listener failed"))
+			sd.conn = nil
+		}
+	}
+
+	return nil
+}
+
+// IsConnected indicates if there is an active connection
+func (sd *SignerDialerEndpoint) IsConnected() bool {
+	sd.mtx.Lock()
+	defer sd.mtx.Unlock()
+	return sd.isConnected()
+}
+
+// IsConnected indicates if there is an active connection
+func (sd *SignerDialerEndpoint) isConnected() bool {
+	return sd.IsRunning() && sd.conn != nil
+}
+
+func (sd *SignerDialerEndpoint) readMessage() (msg RemoteSignerMsg, err error) {
```

Comment:
I think I've seen this before. Is it identical to `readMsg`?

Reply:
yes, now that things got refactored they are almost the same. 
There are a tiny differences with respect to dropping.
I left this behind but I guess I can unify them now

Reply:
if you look at dialer/listener endpoints, I tried to follow almost the same order. they are quite similar.
As usual, there is a balance between unifying them or keeping them separate in case there is a need for something slightly different.

Reply:
I had a look, unifying read/write/request  right now would mean some additional refactoring. Should we open another smaller issue/pr for this?

Reply:
Yes, let's do this in a follow-up issue.

Reply:
Created a follow-up for this issue here:
https://github.com/tendermint/tendermint/issues/3601